File: 0101-0200/101-110/104E_Maximum_Depth_of_Binary_Tree/bfs.js
// 104. Maximum Depth of Binary Tree
// https://leetcode.com/problems/maximum-depth-of-binary-tree/
// T.C.: O(n)
// S.C.: O(n)
const { TreeNode, createTree, printTree } = require('../../../_utils/tree');
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
function maxDepth(root) {
  if (!root) return 0;

  let maxDepth = 0;
  const queue = [];
  queue.push([root, 0]);
  while (queue.length) {
    const [node, depth] = queue.shift();
    maxDepth = Math.max(maxDepth, depth + 1);
    if (node.left) {
      queue.push([node.left, depth + 1]);
    }
    if (node.right) {
      queue.push([node.right, depth + 1]);
    }
  }
  return maxDepth;
}

var root = createTree([3, 9, 20, null, null, 15, 7]);
var expect = 3;
var depth = maxDepth(root);
console.log(depth, expect, depth === expect);

var root = createTree([1, null, 2]);
var expect = 2;
var depth = maxDepth(root);
console.log(depth, expect, depth === expect);

File: 0101-0200/101-110/104E_Maximum_Depth_of_Binary_Tree/cs/bfs.cs
// 104. Maximum Depth of Binary Tree
// https://leetcode.com/problems/maximum-depth-of-binary-tree/
// T.C.: O(n)
// S.C.: O(n)
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class TreeNode
{
  public int val;
  public TreeNode left;
  public TreeNode right;
  public TreeNode(int val = 0, TreeNode left = null, TreeNode right = null)
  {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

public class Solution
{
  public int MaxDepth(TreeNode root)
  {
    if (root == null)
    {
      return 0;
    }

    Queue<Tuple<TreeNode, int>> queue = new Queue<Tuple<TreeNode, int>>();
    queue.Enqueue(new Tuple<TreeNode, int>(root, 0));

    int maxDepth = 0;
    while (queue.Count > 0)
    {
      var (node, depth) = queue.Dequeue();
      var nextDepth = depth + 1;
      maxDepth = Math.Max(maxDepth, nextDepth);

      if (node.left != null)
      {
        queue.Enqueue(new Tuple<TreeNode, int>(node.left, nextDepth));
      }

      if (node.right != null)
      {
        queue.Enqueue(new Tuple<TreeNode, int>(node.right, nextDepth));
      }
    }

    return maxDepth;
  }
}

// [3, 9, 20, null, null, 15, 7]
var root = new TreeNode(3);
root.left = new TreeNode(9);
root.right = new TreeNode(20);
root.right.left = new TreeNode(15);
root.right.right = new TreeNode(7);
var expect = 3;
var result = new Solution().MaxDepth(root);
Console.WriteLine($"{result}, {result == expect}");

// [1, null, 2]
root = new TreeNode(1);
root.right = new TreeNode(2);
expect = 2;
result = new Solution().MaxDepth(root);
Console.WriteLine($"{result}, {result == expect}");
